### PR TITLE
Introduce a new block type

### PR DESCRIPTION
**Pull request by georgeee**
Introduce a block type which has separate sections of body and header.
Use this type everywhere, making necessary conversions to the old type.

Explain how you tested your changes: existing tests cover this refactoring well.

Checklist:

- [x] Modified the current draft of release notes with details on what is completed or incomplete within this project
- [x] Document code purpose, how to use it
  - Mention expected invariants, implicit constraints
- [x] Tests were added for the new behavior
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [x] All tests pass (CI will check this if you didn't)
- [x] Serialized types are in stable-versioned modules
- [x] Does this close issues? None

### CODE REVIEW

File: src/lib/mina_transition/external_transition.ml [hunk 1]
```diff
@@ -80,25 +82,40 @@ module Raw_versioned__ = struct
   end]
 end
 
-include Allocation_functor.Make.Versioned_v1.Sexp (struct
-  let id = "external_transition"
-
-  include Raw_versioned__
-end)
+module Raw = struct
```
Comment (nholland94) [inline src/lib/mina_transition/external_transition.ml:85]: I think the version ppx will be unhappy about this change.

File: src/lib/mina_transition/external_transition.ml [hunk 2]
```diff
@@ -392,7 +436,7 @@ module Validation = struct
        , 'staged_ledger_diff
        , 'protocol_versions )
        with_transition =
-    external_transition State_hash.With_state_hashes.t
+    Block.with_hash
```
Comment (nholland94) [inline src/lib/mina_transition/external_transition.ml:439]: I would prefer to move this validation logic elsewhere, and to patch it up as I did in my original PR. The old validation implementation is actually wrong in a number of ways, which is why I had re-implemented a new validation module and ported over the old validation code to it.

**Review by nholland94 — COMMENTED** [review]
I would prefer to take an approach where we replace `external_transition` et al with a new layer, rather than shoe-horning our new block abstraction into the old layer. Curious what your thoughts are on this.

Otherwise, the changes overall look good to me. Just not sure about putting the new block code into the outdated `mina_transition` library and folding the new format into the existing `External_transition` module.

File: src/lib/mina_networking/mina_networking.ml [hunk 3]
```diff
@@ -26,6 +26,12 @@ type Structured_log_events.t +=
   [@@deriving
     register_event { msg = "Broadcasting snark pool diff over gossip net" }]
 
+(* Hack to circumvent version type ci diff check *)
+module External_transition = struct
```
[the remark this answers is not in the document]
Comment (georgeee) [reply src/lib/mina_networking/mina_networking.ml:30]: Ok, I will drop commit with ppx version changes

I also think this is the right way to go

**Review by nholland94 — APPROVED** [review]

**Pull request merged**